### PR TITLE
Blog post: Make virtus dry to speed up API

### PR DESCRIPTION
**Pull request by ArturT**

### CODE REVIEW

File: _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md [hunk 1]
```diff
@@ -0,0 +1,398 @@
+---
+layout: post
+title:  "Make virtus dry to speed up API"
+date:   2017-02-13 08:00:00 +0200
+author: "Artur Trzop"
+tags: ruby virtus dry-rb
+---
+
+[Virtus](https://github.com/solnic/virtus) is a popular ruby gem described as attributes on steroids for plain old ruby objects. I saw it being used in many API applications so
```
Comment (natif) [inline _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md]: ruby -> Ruby
old ruby - Ruby
API applications - > maybe better applications' APIs or API based applications

File: _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+when I was building API for my app I chose the virtus too. Over a year ago the [virtus gem was abandoned by its creator](https://www.reddit.com/r/ruby/comments/3sjb24/virtus_to_be_abandoned_by_its_creator/) and I learned more about [dry-rb](http://dry-rb.org) - a collection of next-generation Ruby libraries. One of the [virtus problems was performance](https://github.com/solnic/virtus/issues/287) and this article will be about it.
```
Comment (natif) [inline _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md]: virtus -> Virtus

File: _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+
+<img src="/images/blog/posts/make-virtus-dry-to-speed-up-api/dry-virtus.jpg" style="float:right;margin-left:20px;width:400px;" alt="dry virtus" />
+
+# Virtus performance issues
+
+In 2015 I started building API for my gem knapsack_pro in order to [optimize test suite split across many CI nodes](https://knapsackpro.com). Recently I started seeing some significant difference with API performance for large users' test suites. The API response took ~500ms and sometimes event much more.
```
Comment (natif) [inline _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md]: Recently, - missing coma.
and sometimes event -> and some time even

File: _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+
+{% highlight plain %}
+                                             ┃   Mean ┃ StdDev ┃    Min ┃    Max ┃    95 %tile
+━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
+API::V1::BuildDistributionsController#subset ┃  219ms ┃  158ms ┃    0ms ┃  612ms ┃  16ms-529ms
+{% endhighlight %}
+
+I did code profiling of the controller action with the [ruby-prof gem](https://github.com/ruby-prof/ruby-prof) and it turned out that half of the time is spent in virtus gem. Why was it a problem in my case?
+
+The [knapsack_pro gem](https://knapsackpro.com) sends information about test files to API and on the API side the each test file is a separate virtus object. So when someone has large test suite then the API slows down.
```
Comment (natif) [inline _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md]: side the each test - side each test

Comment (natif) [reply _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md]: So -> In result, when

File: _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+
+# Virtus with array of many value objects
+
+Here is the example how I represent the test files in my codebase. Basically, I have a `Node` which is one of the CI nodes where is part of test suite executed. The `Node` has many test files as [value objects](https://en.wikipedia.org/wiki/Value_object).
```
Comment (natif) [inline _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md]: here is an example

File: _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+    attribute :node_index, Integer
+    # here is the performance problem when there is too many test files
+    attribute :test_files, Array[TestFileValue]
+  end
+end
+{% endhighlight %}
+
+# What are options to improve speed?
+
+I was wondering about a reasonable solution for my virtus performance issue. One of it that occurred was to switch to dry-rb but it would require more work to adjust the whole codebase to it. I decided to try to replace virtus value objects with [dry-struct](http://dry-rb.org/gems/dry-struct/) which is a gem built on top of [dry-types](http://dry-rb.org/gems/dry-types/) which provides virtus-like DSL for defining typed struct classes. One of nice virtus feature was input parameter sanitization and coercion. Dry-types would provide that as well.
```
Comment (natif) [inline _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md]: Dry-types would provide -> Dry-types provides

File: _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+  context 'when virtus model with value objects in items' do
+    let(:node_class) { NodeWithVirtusValueObject }
+    let(:item_class) { VirtusItem }
+
+    it_behaves_like 'node'
+  end
+end
+{% endhighlight %}
+
+Now when I had basic test coverage I could add dry libs.
```
Comment (natif) [inline _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md]: I had a basic

File: _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md (hunk 3 above)
Comment (Tomabana) [inline _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md]: my gem knapsack_pro - I'll add link to this gem. I know this link is added in next sentence, but we should add it for the first occurrence.

File: _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+API::V1::BuildDistributionsController#subset ┃  219ms ┃  158ms ┃    0ms ┃  612ms ┃  16ms-529ms
+{% endhighlight %}
+
+I did code profiling of the controller action with the [ruby-prof gem](https://github.com/ruby-prof/ruby-prof) and it turned out that half of the time is spent in virtus gem. Why was it a problem in my case?
+
+The [knapsack_pro gem](https://knapsackpro.com) sends information about test files to API and on the API side each test file is a separate virtus object. In result, when someone has large test suite then the API slows down.
+
+# Virtus with array of many value objects
+
+Here is an example how I represent the test files in my codebase. Basically, I have a `Node` which is one of the CI nodes where is part of test suite executed. The `Node` has many test files as [value objects](https://en.wikipedia.org/wiki/Value_object).
```
Comment (Tomabana) [inline _posts/2017-02-13-make-virtus-dry-to-speed-up-api.md]: where a part of test is executed

**Review by natif — APPROVED** [review]

**Review by Tomabana — APPROVED** [review]

**Pull request merged**